Introduce new `mergeable` requirement

brndnmtthws opened the pull request:
Introduce new `mergeable` requirement, in similar vein to the `approved`
requirement.

Addresses #43.

lkysow commented:
@brndnmtthws I'm going to merge #386 to regen all the mocks and update pegomock. Can you then rebase off master please? Thanks!

brndnmtthws commented:
👍

lkysow commented:
Which VCS providers did you manually test this with? I can help you with testing, I just want to know.

brndnmtthws commented:
Just GitHub. However with all the code shuffling it would be a good idea for me to test it again before you merge this PR.

brndnmtthws commented:
I've rebased on master and updated the PR. Also squashed all the commits into one.

It looks like there are some style changes from my linter. I'm guessing that's because we're using a slightly different version. I can go through and drop those changes if necessary.

Let me test this with GitHub and I'll report back.

lkysow commented on server/events/project_command_builder_test.go:
```diff
@@ -350,6 +350,58 @@ projects:
 			ExpDir:       ".",
 		},
+		{
+			Description: "atlantis.yaml with projectname",
```
The `Description` should be unique between test cases. It gets used in `t.Run` to identify the name of the test. If a test fails, you look back up to find a test that matches that description. This is a duplicate.

You could say something like `atlantis.yaml with mergeable apply requirement`

lkysow commented on server/events/project_command_runner.go:
```diff
@@ -213,8 +215,14 @@ func (p *DefaultProjectCommandRunner) doApply(ctx models.ProjectCommandContext)
 	// Instead the project_command_builder should figure this out and store this information in the ctx. # refactor
-	if p.RequireApprovalOverride {
-		applyRequirements = []string{raw.ApprovedApplyRequirement}
+	if p.RequireMergeableOverride || p.RequireApprovalOverride {
```
Let's move this up and combine it with our check of `ctx.ProjectConfig`. Also if users specify both flags then we should use both requirements:

```go
	// Figure out what our apply requirements are.
	var applyRequirements []string
	if p.RequireApprovalOverride || p.RequireMergeableOverride {
		// If any server flags are set, they override project config.
		if p.RequireApprovalOverride {
			applyRequirements = append(applyRequirements, raw.ApprovedApplyRequirement)
		}
		if p.RequireMergeableOverride {
			applyRequirements = append(applyRequirements, raw.MergeableApplyRequirement)
		}
	} else if ctx.ProjectConfig != nil {
		// Else we use the project config if it's set.
		applyRequirements = ctx.ProjectConfig.ApplyRequirements
	}
```

brndnmtthws replied:
Got it, will do.

brndnmtthws commented:
Everything works, tested with GitHub:

<img width="819" alt="screen shot 2018-12-11 at 1 33 13 pm" src="https://user-images.githubusercontent.com/3129093/49821821-5e387b80-fd49-11e8-9e12-2ca3545f6890.png">

However there's an issue I've noticed, and I'd like to ask for guidance on: should the CLI override options remove any per-project options? Or should it be additive?

Specifically, I'm referring to the logic here: https://github.com/brndnmtthws/atlantis/blob/2a6b2f50edb29545efda24a047ac71214e4a7680/server/events/project_command_runner.go#L218-L223

Currently the logic is additive, so it appends these requirements to the existing list. The downside of this is that it may check the same requirements multiple times if you specify it both per project, and globally. Should the global options override the project options?

Thoughts?

lkysow commented on server/events/vcs/bitbucketcloud/client.go:
```diff
@@ -124,6 +124,27 @@ func (b *Client) PullIsApproved(repo models.Repo, pull models.PullRequest) (bool
+	if err := validator.New().Struct(conflictStatus); err != nil {
+		return false, errors.Wrapf(err, "API response %q was missing fields", string(resp))
+	}
+	if (conflictStatus.MergeImpossible == nil || !*conflictStatus.MergeImpossible) && (conflictStatus.IsConflicted == nil || !*conflictStatus.IsConflicted) {
```
You don't need to check nil's here thanks to the validator

brndnmtthws replied:
Hmm, sure about this? Does a JSON `null` get converted to `false`? I'm not sure exactly how the JSON parser works.

lkysow replied:
The parser will set the field to `nil`, however we're performing additional validation via the `validator` library. When you set the struct tags up, it will require those fields to not be `nil`.

lkysow commented on server/events/vcs/bitbucketcloud/models.go:
```diff
@@ -80,3 +80,8 @@ type Comment struct {
 }
+
+type ConflictStatus struct {
+	MergeImpossible *bool `json:"mergeimpossible,omitempty"`
```
You need to add the validate struct tags here:

```
type ConflictStatus struct {
	MergeImpossible *bool `json:"mergeimpossible,omitempty" validate:"required"`
	IsConflicted    *bool `json:"isconflicted,omitempty" validate:"required"`
}
```

lkysow commented on server/events/vcs/bitbucketserver/client.go:
```diff
@@ -178,6 +178,30 @@ func (b *Client) PullIsApproved(repo models.Repo, pull models.PullRequest) (bool
 }
 
+// PullIsMergeable returns true if the merge request has no conflicts and can be merged.
+func (b *Client) PullIsMergeable(repo models.Repo, pull models.PullRequest) (bool, error) {
```
Based on https://community.atlassian.com/t5/Bitbucket-questions/Best-way-to-automatically-detect-when-a-pull-request-resulted-in/qaq-p/402120

I don't think this works for bitbucket server.

Maybe leave it out for now until I have time to test it.

brndnmtthws replied:
I don't know either, however I found this which suggests it might work:

https://community.atlassian.com/t5/Bitbucket-questions/Best-way-to-automatically-detect-when-a-pull-request-resulted-in/qaq-p/402120#M28185

Although it looks like that's referring to their cloud offering, IDK.

brndnmtthws replied:
I changed this to just `return true, nil` for now. I left the code in there (commented out), but I can remove it if you prefer.

lkysow commented on server/events/vcs/github_client.go:
```diff
@@ -121,6 +121,15 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
 
+// PullIsMergeable returns true if the pull request is mergeable.
+func (g *GithubClient) PullIsMergeable(repo models.Repo, pull models.PullRequest) (bool, error) {
+	githubPr, err := g.GetPullRequest(repo, pull.Num)
```
`githubPR` for idiomatic naming please

lkysow commented on server/events/yaml/parser_validator_test.go:
```diff
@@ -226,6 +226,74 @@ workflows:
 			},
 		},
+		{
+			description: "project field with autoplan",
```
these descriptions should be different

lkysow commented on server/server.go:
```diff
@@ -102,7 +102,10 @@ type UserConfig struct {
-	RequireApproval        bool            `mapstructure:"require-approval"`
+	RequireApproval bool `mapstructure:"require-approval"`
+	// RequireMergeable is similar to RequireApproval, but won't allow running
+	// `terraform apply` unless the PR is actually mergable.
```
Change to:
```
// RequireMergeable is whether to require pull requests to be mergeable before
// allowing terraform apply's to run.
```

I prefer not having to refer to another struct field's docs if I can help it, need to fix mispelling of mergeable, and I try to avoid the backticks since they're not parsed by godoc

lkysow commented on testdrive/github.go:
```diff
@@ -56,7 +56,6 @@ func (g *Client) CheckForkSuccess(ownerName string, forkRepoName string) bool {
 // CreateWebhook creates a GitHub webhook to send requests to our local ngrok.
 func (g *Client) CreateWebhook(ownerName string, repoName string, hookURL string) error {
 	atlantisHook := &github.Hook{
-		Name:   github.String("web"),
```
why is this removed?

brndnmtthws replied:
See here:

https://github.com/google/go-github/pull/1032

lkysow commented:
To answer your question, the flags should override the yaml settings. When server-side config is released, the flags will be deprecated.

brndnmtthws commented:
I think everything has been addressed, but you should probably take another look.

lkysow commented:
I managed to test Bitbucket Server. We need to hit another endpoint and the response is different:
```
curl 'http://localhost:7990/rest/api/1.0/projects/AT/repos/atlantis-example/pull-requests/1/merge'
{"canMerge":true,"conflicted":false,"outcome":"CLEAN","vetoes":[]}%
```

lkysow commented on cmd/server.go:
```diff
@@ -20,7 +20,7 @@ import (
 	"strings"
 
-	"github.com/mitchellh/go-homedir"
+	homedir "github.com/mitchellh/go-homedir"
```
Why do you have all these named imports? What tool is adding them back?

lkysow replied:
Can you remove them please unless there is a requirement somewhere that we need them now.l

brndnmtthws replied:
I assumed it was the linter, I didn't add them explicitly. I'm just using `make gometalint`.

lkysow replied:
The linter won't add anything, just fail

brndnmtthws replied:
Actually, it seems like the issue was using `goreturns` to format. I switched to using `goimports` and it stopped doing it.

lkysow commented on cmd/server.go:
```diff
@@ -186,6 +187,11 @@ var boolFlags = []boolFlag{
 		description:  "Require pull requests to be \"Approved\" before allowing the apply command to be run.",
 		defaultValue: false,
 	},
+	{
```
There are a number of tests in `server_test.go` that need to be updated.

lkysow commented on server/events/project_command_runner_test.go:
```diff
@@ -275,7 +294,43 @@ func TestDefaultProjectCommandRunner_Apply(t *testing.T) {
 				},
 			},
-			expSteps: []string{"approved", "apply"},
+			expSteps: []string{"apply"},
```
the steps still need to be fixed here

lkysow commented on server/events/yaml/raw/project_test.go:
```diff
@@ -101,16 +101,32 @@ func TestProject_Validate(t *testing.T) {
 		},
 		{
-			description: "apply reqs with valid",
+			description: "apply reqs with approved requiremed",
```
typo

brndnmtthws commented:
Something like this?: https://gist.github.com/brndnmtthws/d828c30fea582504c686845fb1b51cfa

I'm not quite sure how to handle `vetoes` correctly in the JSON struct.

lkysow commented:
> Something like this?: https://gist.github.com/brndnmtthws/d828c30fea582504c686845fb1b51cfa
> 
> I'm not quite sure how to handle `vetoes` correctly in the JSON struct.

Patch looks great. Let's ignore vetoes, dunno either 😆

brndnmtthws commented:
I think that's everything addressed, again.

lkysow commented on server/events/project_command_runner_test.go:
```diff
@@ -385,6 +442,7 @@ func TestDefaultProjectCommandRunner_Apply(t *testing.T) {
 			When(mockApply.Run(ctx, nil, repoDir)).ThenReturn("apply", nil)
 			When(mockRun.Run(ctx, nil, repoDir)).ThenReturn("run", nil)
 			When(mockApproved.PullIsApproved(ctx.BaseRepo, ctx.Pull)).ThenReturn(true, nil)
+			When(mockMergeable.PullIsMergeable(ctx.BaseRepo, ctx.Pull)).ThenReturn(true, nil)
```
you're missing adding the case statement below to verify your mock was called

lkysow commented on server/events/vcs/bitbucketserver/client.go:
```diff
@@ -178,6 +178,30 @@ func (b *Client) PullIsApproved(repo models.Repo, pull models.PullRequest) (bool
+	if err != nil {
+		return false, err
+	}
+	var conflictStatus ConflictStatus
```
I think `MergeStatus` is a better name

lkysow commented on server/events/vcs/bitbucketserver/models.go:
```diff
@@ -64,3 +64,10 @@ type Changes struct {
+type ConflictStatus struct {
+	CanMerge   *bool    `json:"canMerge,omitempty" validate:"required"`
+	Conflicted *bool    `json:"conflicted,omitempty" validate:"required"`
+	Outcome    *bool    `json:"outcome,omitempty" validate:"required"`
```
We don't care about the `Outcome` and `Vetoes` fields so you can omit them from the struct. The json will parse fine.

lkysow commented:
Thanks for the awesome PR @brndnmtthws. I made some additional changes in #388 specifically around checking another field for github.
